### PR TITLE
Add extract_assignment assist

### PR DESCRIPTION
**Pull request by Jesse-Bakker**
Add extract-assignment assist (#7006).

Assist is for now only implemented on if/match-statements where the assigment is the last statement in every block,
as for other cases, one would have to check whether the assignment has effects on the rest of the block and
extract a temporary variable for it in the block.

### CODE REVIEW

File: crates/assists/src/handlers/extract_assignment.rs [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+        },
+    )
+}
+
+fn exprify_if(statement: &IfExpr, name: &str) -> Option<Expr> {
+    let then_branch = exprify_block(statement.then_branch()?, name)?;
+    let else_branch = match statement.else_branch()? {
+        ElseBranch::Block(block) => ElseBranch::Block(exprify_block(block, name)?),
+        ElseBranch::IfExpr(expr) => {
+            ElseBranch::IfExpr(IfExpr::cast(exprify_if(&expr, name)?.syntax().to_owned())?)
```
Comment (Veykril) [inline crates/assists/src/handlers/extract_assignment.rs:62]: A test for this chained `else if` part would be nice I think, could also make use of a [mark](https://github.com/rust-analyzer/rust-analyzer/blob/master/crates/test_utils/src/mark.rs).

File: crates/assists/src/handlers/extract_assignment.rs [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+    }
+
+    let mut stmts: Vec<_> = block.statements().collect();
+    let stmt = stmts.pop()?;
+
+    if let Stmt::ExprStmt(stmt) = stmt {
+        if let Expr::BinExpr(expr) = stmt.expr()? {
+            if expr.op_kind()? == BinOp::Assignment && expr.lhs()?.name_ref()?.to_string() == name {
+                // The last statement in the block is an assignment to the name we want
+                return Some(make::block_expr(stmts, Some(expr.rhs()?)));
```
Comment (Veykril) [inline crates/assists/src/handlers/extract_assignment.rs:80]: A small test to see that it doesn't discard the other statements in a block that might exist would also be nice.

File: crates/assists/src/handlers/extract_assignment.rs [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+}"#,
+            r#"
+fn foo() {
+    let mut a = 1;
+
+    a = if true {
+        2
+    } else {
+        3
+    }
```
[the remark this answers is not in the document]
Comment (Jesse-Bakker) [reply crates/assists/src/handlers/extract_assignment.rs:115]: Yep

**Comment by Jesse-Bakker** [comment]
Added support for match. Getting correct indentation is a bit of a PITA and I'm not sure if I did it correctly

File: crates/assists/src/handlers/extract_assignment.rs [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+//
+//     foo = if true {
+//         5
+//     } else {
+//         4
+//     };
+// }
+// ```
+pub(crate) fn extract_assigment(acc: &mut Assists, ctx: &AssistContext) -> Option<()> {
+    let name = ctx.find_node_at_offset::<ast::NameRef>()?.to_string();
```
Comment (Veykril) [inline crates/assists/src/handlers/extract_assignment.rs:40]:
```suggestion
    let name = ctx.find_node_at_offset::<ast::NameRef>()?.as_name();
```
With this and the comment below we can get rid of the string usage and compare by `Name`s I believe. The method there comes from [AsName](https://github.com/rust-analyzer/rust-analyzer/blob/master/crates/hir_expand/src/name.rs#L75)

File: crates/assists/src/handlers/extract_assignment.rs [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+        return None;
+    }
+
+    let mut stmts: Vec<_> = block.statements().collect();
+    let stmt = stmts.pop()?;
+
+    if let ast::Stmt::ExprStmt(stmt) = stmt {
+        if let ast::Expr::BinExpr(expr) = stmt.expr()? {
+            if expr.op_kind()? == ast::BinOp::Assignment
+                && expr.lhs()?.name_ref()?.to_string() == name
```
Comment (Veykril) [inline crates/assists/src/handlers/extract_assignment.rs:109]:
```suggestion
                && expr.lhs()?.name_ref()?.as_name() == name
```
See comment above

**Comment by Veykril** [comment]
Ye indentation is an unsolved problem still iirc. Ideally we would be able to just format the produced syntax but thats yet to be done.

LGTM overall, just a small nit.
bors d+

**Comment by matklad** [comment]
See https://github.com/rust-analyzer/rust-analyzer/issues/4182 for indent.

**Pull request merged**